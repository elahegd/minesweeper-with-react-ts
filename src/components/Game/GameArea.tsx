import React, { FC, ReactNode } from 'react';
import styled from "@emotion/styled";

export interface GameAreaProps {
    children: ReactNode;
}

export const GameArea:FC<GameAreaProps> = ({ children }) => {
  return (
    <Frame>{children}</Frame>
  )
}

const Frame = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    border: 6px solid #e3e3e3;
    background-color: #e3e3e3;
    width: fit-content;
    margin: 0 auto;
    position: relative;
`;
