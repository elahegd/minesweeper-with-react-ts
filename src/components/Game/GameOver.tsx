import { FC, memo } from 'react';
import styled from '@emotion/styled';

export interface GameOverProps extends FrameProps {
    onClick: () => void;
}

export const GameOver:FC<GameOverProps> = memo(({ onClick, isWin }) => {
  return (
    <Frame onClick={ onClick } isWin={isWin}>
        {isWin ? '😎' : '🙁'}
    </Frame>
  )
});

// Stryker disable next-line StringLiteral
GameOver.displayName = "GameOver";

interface FrameProps {
    isWin: boolean;
}

const Frame = styled.div<FrameProps>`
    top: 50%;
    left: 50%;
    z-index: 11;
    width: 8vw;
    height: 8vw;
    font-size: 4vw;
    text-align: center;
    user-select: none;
    cursor: pointer;
    line-height: 8vw;
    position: absolute;
    border-radius: 50%;
    pointer-events: auto;
    background-color: #d1d1d1;
    transform: translate(-50%, -50%);
    border: 1px solid rgba(51, 51, 51, 0.25);
`;
