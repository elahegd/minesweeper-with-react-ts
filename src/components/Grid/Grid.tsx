import { FC, memo } from 'react';

import styled from "@emotion/styled";

import { Coords, Field } from "../../helpers/Field";

import Cell from "./Cell";

export interface GridProps {
    children: Field;
    onClick: (coords: Coords) => void;
    onContextMenu: (coords: Coords) => void;
}

export const Grid: FC<GridProps> = memo(({ children, ...rest}) => {
    return (
      <Wrapped size={children.length}>{children.map((row, y) => row.map((cell, x) => (<Cell
          key={x}
          coords={[y, x]} 
          {...rest}
        >
              {cell}
          </Cell>))
          
      )}</Wrapped>
    )
});
// Stryker disable next-line StringLiteral
Grid.displayName = "Grid";

interface WrapperSize {
    size: number;
}

const Wrapped = styled.div<WrapperSize>`
    display: grid;
    grid-template-columns: repeat(${({size}) => size}, auto);
    width: max-content;
    padding: 1vw;
`;