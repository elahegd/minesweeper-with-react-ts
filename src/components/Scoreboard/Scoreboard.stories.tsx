import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Scoreboard, ScoreboardProps } from './Scoreboard';

export default {
    title: "Scoreboard/Scoreboard",
    component: Scoreboard
} as Meta;

const Template: Story<ScoreboardProps> = (args) => <Scoreboard {...args} />;

export const ScoreboardExample = Template.bind({});
ScoreboardExample.args = {
    timer: "000",
    levels: ['beginner', 'intermediat', 'expert'],
    defaultValue: 'beginner',
    bombs: "010",
}