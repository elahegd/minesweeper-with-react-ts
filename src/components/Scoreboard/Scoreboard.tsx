import React, { FC, ChangeEvent } from 'react';
import styled from "@emotion/styled";

import { Reset } from "./Reset";
import { Counter } from "./Counter";
import { Level } from './Level';

export interface ScoreboardProps {
    timer: string;
    levels: string[];
    defaultValue?: string;
    onChangeLevel: (event: ChangeEvent<HTMLSelectElement>) => void;
    onReset: () => void;
    bombs: string;
}

export const Scoreboard: FC<ScoreboardProps> = ({timer, levels, defaultValue, onChangeLevel, onReset, bombs}) => {
  return (
    <Wrapper>
      <Counter>{timer}</Counter>
      <ResetAndLevelWrapper>
        <Level value={defaultValue} onChange={onChangeLevel}>
          {levels}
        </Level>
        <Reset onReset={onReset} />
      </ResetAndLevelWrapper>
      <Counter>{bombs}</Counter>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;
const ResetAndLevelWrapper = styled.div`
  display: flex;
`;