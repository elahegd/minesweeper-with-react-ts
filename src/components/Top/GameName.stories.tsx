import React from 'react';
import { Story, Meta } from '@storybook/react';

import { GameName } from './GameName';

export default {
    title: "Top/GameName",
    component: GameName
} as Meta;

const Template: Story = (args) => <GameName {...args} />;

export const MinesweeperGameName = Template.bind({});
MinesweeperGameName.args = {
    children: 'minesweeper'
}