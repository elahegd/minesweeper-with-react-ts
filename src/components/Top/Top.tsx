import { FC, memo } from 'react';
import styled from '@emotion/styled';

import { Legend, LegendProps } from './Legend';
import { GameName } from './GameName';

export interface TopProps extends LegendProps {
    children: string;
}

export const Top: FC<TopProps> = memo(({ children, ...legendProps }) => (
  <Header>
    <GameName>{children}</GameName>
    <Legend {...legendProps} />
  </Header>
));
// Stryker disable next-line StringLiteral
// Useful for debuging in devtools
Top.displayName = 'Top';

const Header = styled.header`
  text-align: center;
  position: relative;
  display: inline-block;
`;