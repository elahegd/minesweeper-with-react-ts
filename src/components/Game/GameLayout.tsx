import React, { FC, ReactNode } from 'react';

import { GameArea } from './GameArea';
import { Wrapper } from './Wrapper';

export interface GameLayoutProps {
    top: ReactNode;
    children: ReactNode;
}

export const GameLayout:FC<GameLayoutProps> = ({ top, children }) => {
  return (
    <Wrapper>
        {top}
        <GameArea>{children}</GameArea>
    </Wrapper>
  )
}
