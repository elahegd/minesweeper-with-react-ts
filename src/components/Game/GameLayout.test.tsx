import React from 'react';
import { render } from "@testing-library/react";

import {GameLayout} from "./GameLayout";
import { GameArea } from "./GameArea";
import { GameOver } from './GameOver';

import { Top } from "../Top";
import { Scoreboard } from "../Scoreboard";
import { Grid } from "../Grid";

import { Field } from "../../helpers/Field";
import { fieldGenerator } from "../../helpers/__mocks__/Field";

describe("Render GameLayout component with props", () => {
    it("GameLayout renders correctly", () => {
        const { asFragment } = render(
        <GameLayout
            top= { 
                <Top 
                    feature={'Flag'}
                    firstAction={'ctrl'}
                    secondAction={'click'}>
                minesweeper
                </Top>
            }
            >
                <>
                    <GameArea>
                    <Scoreboard 
                        timer="000"
                        bombs="000"
                        levels={['beginner', 'intermediate', 'expert']}
                        defaultValue='beginner'
                        onReset={() => null}
                        onChangeLevel={() => null}
                    />
                        <GameOver onClick={() => null} isWin={true} />
                        <Grid onClick={() => null} onContextMenu={() => null}>
                            {fieldGenerator(9) as Field}
                        </Grid>
                    </GameArea>
                </>
            </GameLayout>);

        expect(asFragment()).toMatchSnapshot();;
    });
});