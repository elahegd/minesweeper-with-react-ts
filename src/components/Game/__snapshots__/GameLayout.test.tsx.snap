// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render GameLayout component with props GameLayout renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-129n4x9"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-12ap3z0"
        >
          <span
            class="css-udfepp"
          >
            ctrl
          </span>
          <span
            class="css-9c23k1"
          >
             + click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="css-hajvvz"
    >
      <div
        class="css-hajvvz"
      >
        <div
          class="css-1n7y7g1"
        >
          <div
            class="css-d2q1pa"
          >
            000
          </div>
          <div
            class="css-ov1ktg"
          >
            <select
              class="css-1lpmfg0"
            >
              <option
                class="css-zcmdxr"
                value="beginner"
              >
                beginner
              </option>
              <option
                class="css-zcmdxr"
                value="intermediate"
              >
                intermediate
              </option>
              <option
                class="css-zcmdxr"
                value="expert"
              >
                expert
              </option>
            </select>
            <div>
              <button
                class="css-1kb3z2h"
              >
                ðŸ˜€
              </button>
            </div>
          </div>
          <div
            class="css-d2q1pa"
          >
            000
          </div>
        </div>
        <div
          class="css-ap4m5y"
        >
          ðŸ˜Ž
        </div>
        <div
          class="css-1a9cl3f"
          size="9"
        >
          <div
            class="css-1l2qare"
            data-testid="0,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="0,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="0,2"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="0,3"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="0,4"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="0,5"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="0,6"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="0,7"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_0,7"
            />
          </div>
          <div
            class="css-y5m8ng"
            data-testid="0,8"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1l2qare"
            data-testid="1,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="1,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="1,2"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="1,3"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="1,4"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="1,5"
            role="cell"
          >
            1
          </div>
          <div
            class="css-13kl6wr"
            data-testid="1,6"
            role="cell"
          >
            2
          </div>
          <div
            class="css-13kl6wr"
            data-testid="1,7"
            role="cell"
          >
            2
          </div>
          <div
            class="css-13kl6wr"
            data-testid="1,8"
            role="cell"
          >
            2
          </div>
          <div
            class="css-1l2qare"
            data-testid="2,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="2,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="2,2"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="2,3"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="2,4"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="2,5"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_2,5"
            />
          </div>
          <div
            class="css-y5m8ng"
            data-testid="2,6"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="2,7"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="2,8"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_2,8"
            />
          </div>
          <div
            class="css-1l2qare"
            data-testid="3,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="3,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="3,2"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="3,3"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1ejwm48"
            data-testid="3,4"
            role="cell"
          >
            3
          </div>
          <div
            class="css-13kl6wr"
            data-testid="3,5"
            role="cell"
          >
            2
          </div>
          <div
            class="css-13kl6wr"
            data-testid="3,6"
            role="cell"
          >
            2
          </div>
          <div
            class="css-y5m8ng"
            data-testid="3,7"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="3,8"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1l2qare"
            data-testid="4,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="4,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="4,2"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="4,3"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_4,3"
            />
          </div>
          <div
            class="css-13kl6wr"
            data-testid="4,4"
            role="cell"
          >
            2
          </div>
          <div
            class="css-12dzs03"
            data-testid="4,5"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_4,5"
            />
          </div>
          <div
            class="css-y5m8ng"
            data-testid="4,6"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1l2qare"
            data-testid="4,7"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="4,8"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,0"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,1"
            role="cell"
          >
            1
          </div>
          <div
            class="css-13kl6wr"
            data-testid="5,2"
            role="cell"
          >
            2
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,3"
            role="cell"
          >
            1
          </div>
          <div
            class="css-13kl6wr"
            data-testid="5,4"
            role="cell"
          >
            2
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,5"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,6"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,7"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="5,8"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="6,0"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="6,1"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_6,1"
            />
          </div>
          <div
            class="css-y5m8ng"
            data-testid="6,2"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1l2qare"
            data-testid="6,3"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="6,4"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="6,5"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="6,6"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="6,7"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="6,8"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_6,8"
            />
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,0"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,1"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,2"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,3"
            role="cell"
          >
            1
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,4"
            role="cell"
          >
            1
          </div>
          <div
            class="css-13kl6wr"
            data-testid="7,5"
            role="cell"
          >
            2
          </div>
          <div
            class="css-y5m8ng"
            data-testid="7,6"
            role="cell"
          >
            1
          </div>
          <div
            class="css-1ejwm48"
            data-testid="7,7"
            role="cell"
          >
            3
          </div>
          <div
            class="css-13kl6wr"
            data-testid="7,8"
            role="cell"
          >
            2
          </div>
          <div
            class="css-1l2qare"
            data-testid="8,0"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="8,1"
            role="cell"
          >
            0
          </div>
          <div
            class="css-1l2qare"
            data-testid="8,2"
            role="cell"
          >
            0
          </div>
          <div
            class="css-y5m8ng"
            data-testid="8,3"
            role="cell"
          >
            1
          </div>
          <div
            class="css-12dzs03"
            data-testid="8,4"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_8,4"
            />
          </div>
          <div
            class="css-13kl6wr"
            data-testid="8,5"
            role="cell"
          >
            2
          </div>
          <div
            class="css-12dzs03"
            data-testid="8,6"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_8,6"
            />
          </div>
          <div
            class="css-13kl6wr"
            data-testid="8,7"
            role="cell"
          >
            2
          </div>
          <div
            class="css-12dzs03"
            data-testid="8,8"
            role="cell"
          >
            <div
              class="css-81z0jo"
              data-testid="bomb_8,8"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
